stdlib/public/runtime/ProtocolConformance.cpp

```diff
@@ -505,14 +508,23 @@ swift_conformsToProtocolImpl(const Metadata *const type,
       processSection(section);
   }
 
-  // Try the search again to look for the most specific cached conformance.
-  found = searchInConformanceCache(type, protocol);
+  // Find the most specific conformance that was scanned.
+  const WitnessTable *foundWitness = nullptr;
+  const Metadata *searchType = type;
+  while (!foundWitness && searchType) {
+    foundWitness = foundWitnesses.lookup(searchType);
 
-  // If it's not authoritative, then add an authoritative entry for this type.
-  if (!found.first)
-    C.cacheResult(type, protocol, found.second, snapshot.count());
+    // If there's no entry here, move up to the superclass (if any).
+    if (!foundWitness)
+      searchType = _swift_class_getSuperclass(searchType);
+  }
+
+  // If it's for a superclass or if we didn't find anything, then add an
+  // authoritative entry for this type.
+  if (searchType != type)
```

Comment:
Shouldn't we call `cacheResult` regardless of whether the types match here?  Isn't that the point of `snapshot.count()`?

Reply:
When the types match, then we'll have already cached this result up above during the conformance scan. Caching here would be slightly wasteful, although otherwise harmless.

This whole thing is a little weird. `searchInConformanceCache` checks superclasses, and we don't add cache entries for subclasses when we find a superclass conformance in the cache. Whether a subclass gets its own cache entry depends on whether you looked up a conformance on a superclass first. We should probably figure out the right way to do this (always add cache entries for subclasses, or never) and fix it up, but another time.... For now, this matches the previous behavior.